tests/mocha/comment_test.js

```diff
@@ -133,4 +146,71 @@ suite('Comments', function() {
       assertBubbleSize(this.comment, 100, 100);
     });
   });
+  suite('Restore Comment', function() {
+    setup(function() {
+      this.blockCount = this.workspace.getAllBlocks().length;
+      this.tempBlock = Blockly.Xml.domToBlock(Blockly.Xml.textToDom(
+        '<block type="empty_block"/>'
+      ), this.workspace);
+      this.tempBlock.setCommentText('test temp text');
+    });
+    function assertComment(workspace, text) {
+      const count = workspace.getAllBlocks().length;
+      // show comment
+      const tempBlock = workspace.getAllBlocks()[count - 1];
+      tempBlock.comment.setVisible(true);
+      // check comment bubble size
+      const comment = tempBlock.getCommentIcon();
+      const bubbleSize = comment.getBubbleSize();
+      chai.assert.isNotNaN(bubbleSize.width);
+      chai.assert.isNotNaN(bubbleSize.height);
+      chai.assert.equal(comment.textarea_.value, text);
+    }
+    test('Trashcan', function() {
+      // delete block
+      this.tempBlock.checkAndDelete();
+      chai.assert.equal(this.workspace.getAllBlocks().length, this.blockCount);
+      // open trashcan
+      simulateClick(this.workspace.trashcan.svgGroup_);
+      // place from trashcan
+      simulateClick(this.workspace.trashcan.flyout.svgGroup_.querySelector('.blocklyDraggable'));
+      chai.assert.equal(this.workspace.getAllBlocks().length, this.blockCount + 1);
+      // check comment
+      assertComment(this.workspace, 'test temp text');
+    });
+    test('Undo', function() {
+      // delete block
+      this.tempBlock.checkAndDelete();
+      chai.assert.equal(this.workspace.getAllBlocks().length, this.blockCount);
+      // undo
+      this.workspace.undo(false);
+      chai.assert.equal(this.workspace.getAllBlocks().length, this.blockCount + 1);
+      // check comment
+      assertComment(this.workspace, 'test temp text');
+    });
+    test('Redo', function() {
+      // undo & undo
+      this.workspace.undo(false);
```

Comment:
Thank you so much for adding unit tests :D But the way these are structured needs to be changed a bit. Right now this test is relying on the state of the previous test, which means they can't be run on their own or in parallel.

Since this is more of an integration test (it tests a bunch of systems and how they "integrate", rather than just one "unit") you could make it one large test, rather than breaking it out into separate ones. Or even better, instead of writing an automated test, you could document your test steps in a comment on this PR, and we can add the steps to our manual testing document =)

Reply:
Thank you for your advice.
Which action do you think is better?
1. Remove the added test code and document the test steps in this PR.
2. Separate the added test code into another file (e.g.  comment_restore_test.js) and modify the structure.

Reply:
I gave it a try to separate the test code and modify the structure.